diff --git a/gas/config/tc-aarch64.c b/gas/config/tc-aarch64.c
index a21fbe6cf5e..12eb13594c9 100644
--- a/gas/config/tc-aarch64.c
+++ b/gas/config/tc-aarch64.c
@@ -8930,6 +8930,16 @@ static const struct aarch64_cpu_option_table aarch64_cpus[] = {
 				  | AARCH64_FEATURE_DOTPROD
 				  | AARCH64_FEATURE_PROFILE),
 				  "Neoverse N1"},
+  {"neoverse-n2", AARCH64_FEATURE (AARCH64_ARCH_V8_5,
+				   AARCH64_FEATURE_BFLOAT16
+				 | AARCH64_FEATURE_I8MM
+				 | AARCH64_FEATURE_F16
+				 | AARCH64_FEATURE_SVE
+				 | AARCH64_FEATURE_SVE2
+				 | AARCH64_FEATURE_SVE2_BITPERM
+				 | AARCH64_FEATURE_MEMTAG
+				 | AARCH64_FEATURE_RNG),
+				 "Neoverse N2"},
   {"neoverse-v1", AARCH64_FEATURE (AARCH64_ARCH_V8_4,
 			    AARCH64_FEATURE_PROFILE
 			  | AARCH64_FEATURE_CVADP
diff --git a/gas/doc/c-aarch64.texi b/gas/doc/c-aarch64.texi
index 00eb6fefb32..f5d26695749 100644
--- a/gas/doc/c-aarch64.texi
+++ b/gas/doc/c-aarch64.texi
@@ -72,6 +72,7 @@ on the target processor.  The following processor names are recognized:
 @code{exynos-m1},
 @code{falkor},
 @code{neoverse-n1},
+@code{neoverse-n2},
 @code{neoverse-v1},
 @code{neoverse-e1},
 @code{qdf24xx},
