--- a/gprofng/src/Makefile.am
+++ b/gprofng/src/Makefile.am
@@ -115,11 +115,13 @@ AM_CXXFLAGS = $(AM_CFLAGS)
 QLParser.tab.cc QLParser.tab.hh: QLParser.yy
 	$(BISON) $^
 
+if BUILD_MAN
 man_MANS = gprofng.1 \
 	gp-archive.1 \
 	gp-collect-app.1 \
 	gp-display-src.1 \
 	gp-display-text.1
+endif
 
 MAINTAINERCLEANFILES = $(man_MANS)
 
--- a/gprofng/src/Makefile.in
+++ b/gprofng/src/Makefile.in
@@ -543,11 +543,11 @@ AM_CFLAGS = $(GPROFNG_CFLAGS) $(PTHREAD_CFLAGS) \
 	$(GPROFNG_NO_SWITCH_CFLAGS)
 
 AM_CXXFLAGS = $(AM_CFLAGS)
-man_MANS = gprofng.1 \
-	gp-archive.1 \
-	gp-collect-app.1 \
-	gp-display-src.1 \
-	gp-display-text.1
+@BUILD_MAN_TRUE@man_MANS = gprofng.1 \
+@BUILD_MAN_TRUE@	gp-archive.1 \
+@BUILD_MAN_TRUE@	gp-collect-app.1 \
+@BUILD_MAN_TRUE@	gp-display-src.1 \
+@BUILD_MAN_TRUE@	gp-display-text.1
 
 MAINTAINERCLEANFILES = $(man_MANS)
 BUILT_SOURCES = QLParser.tab.hh
